src/raftstore/store/snap.rs

```diff
@@ -618,6 +670,1025 @@ mod v1 {
+            let meta_tmp_path = dir_path.join(format!("{}{}", meta_filename, TMP_FILE_SUFFIX));
+            let meta_file = MetaFile {
+                path: meta_path,
+                tmp_path: meta_tmp_path,
+                ..Default::default()
+            };
+            let mut s = Snap {
+                display_path: display_path,
+                cf_files: cf_files,
+                cf_index: 0,
+                meta_file: meta_file,
+                size_track: size_track,
+            };
+            // load snapshot meta if meta_file exists
+            if file_exists(&s.meta_file.path) {
+                try!(s.load_snapshot_meta())
+            }
+            Ok(s)
+        }
+
+        pub fn new_for_building<T: Into<PathBuf>>(dir: T,
+                                                  key: &SnapKey,
+                                                  snap: &DbSnapshot,
+                                                  size_track: Arc<RwLock<u64>>)
+                                                  -> RaftStoreResult<Snap> {
+            let mut s = try!(Snap::new(dir, key, size_track, SNAPSHOT_CFS, true));
+            try!(s.init_for_building(snap));
+            Ok(s)
+        }
+
+        pub fn new_for_sending<T: Into<PathBuf>>(dir: T,
+                                                 key: &SnapKey,
+                                                 size_track: Arc<RwLock<u64>>)
+                                                 -> RaftStoreResult<Snap> {
+            let mut s = try!(Snap::new(dir, key, size_track, SNAPSHOT_CFS, true));
+            try!(s.init_for_sending());
+            Ok(s)
+        }
+
+        pub fn new_for_receiving<T: Into<PathBuf>>(dir: T,
+                                                   key: &SnapKey,
+                                                   snapshot_meta: SnapshotMeta,
+                                                   size_track: Arc<RwLock<u64>>)
+                                                   -> RaftStoreResult<Snap> {
+            let cfs = try!(check_snapshot_meta(&snapshot_meta));
+            let mut s = try!(Snap::new(dir, key, size_track, &cfs[..], false));
+            try!(s.init_for_receiving(snapshot_meta));
+            Ok(s)
+        }
+
+        pub fn new_for_applying<T: Into<PathBuf>>(dir: T,
+                                                  key: &SnapKey,
+                                                  size_track: Arc<RwLock<u64>>)
+                                                  -> RaftStoreResult<Snap> {
+            let s = try!(Snap::new(dir, key, size_track, SNAPSHOT_CFS, false));
+            Ok(s)
+        }
+
+        fn init_for_building(&mut self, snap: &DbSnapshot) -> RaftStoreResult<()> {
+            if self.exists() {
+                return Ok(());
+            }
+            let env_opt = EnvOptions::new();
+            for cf_file in &mut self.cf_files {
+                if plain_file_used(cf_file.cf) {
```

Comment:
If there are tmp files of a snapshot, they should be wiped out on `drop` when the program is shut down.
If snapshot files are done being built, they are left in hard drive to be read later.